staging/src/k8s.io/api/flowcontrol/v1alpha1/types.go

```diff
@@ -0,0 +1,296 @@
+	// `type` is the type of flow distinguisher method
+	Type FlowDistinguisherMethodType `json:"type,omitempty" protobuf:"bytes,1,opt,name=type"`
+}
+
+// PriorityLevelConfigurationReference contains information that points to the "request-priority" being used.
+type PriorityLevelConfigurationReference struct {
+	// `name` is the name of resource being referenced
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+}
+
+// PolicyRuleWithSubjects prescribes a test that applies to a request to an apiserver. The test considers the subject
+// making the request, the verb being requested, and the resource to be acted upon.
+type PolicyRuleWithSubjects struct {
+	// `subjects` is the list of normal user, serviceaccount, or group that this rule cares about.
+	// +optional
+	Subjects []Subject `json:"subjects,omitempty" protobuf:"bytes,1,opt,name=subjects"`
+	// `rule` is the target verb, resource or the subresource the rule cares about. APIGroups, Resources, etc.
+	// Required.
+	Rule PolicyRule `json:"rule" protobuf:"bytes,2,opt,name=rule"`
+}
+
+// PolicyRule holds information that describes a policy rule, but does not contain information
+// about who the rule applies to or which namespace the rule applies to.
+type PolicyRule struct {
+	// `verbs` is a list of Verbs that apply to ALL the ResourceKinds and AttributeRestrictions contained in this rule.
+	// VerbAll represents all verbs.
+	Verbs []string `json:"verbs" protobuf:"bytes,1,rep,name=verbs"`
+	// `apiGroups` is the name of the APIGroup that contains the resources.  If multiple API groups are specified, any action requested against one of
+	// the enumerated resources in any API group will be allowed. '*' represents all api groups.
+	// +optional
+	APIGroups []string `json:"apiGroups,omitempty" protobuf:"bytes,2,rep,name=apiGroups"`
+	// `resources` is a list of resources this rule applies to.  ResourceAll represents all resources.
+	// +optional
+	Resources []string `json:"resources,omitempty" protobuf:"bytes,3,rep,name=resources"`
+	// `nonResourceURLs` is a set of partial urls that a user should have access to.  *s are allowed, but only as the full, final step in the path
+	// Since non-resource URLs are not namespaced, this field is only applicable for ClusterRoles referenced from a ClusterRoleBinding.
+	// Rules can either apply to API resources (such as "pods" or "secrets") or non-resource URL paths (such as "/api"),  but not both.
+	// NonResourceAll represents all non-resource urls.
+	// +optional
+	NonResourceURLs []string `json:"nonResourceURLs,omitempty" protobuf:"bytes,6,rep,name=nonResourceURLs"`
+}
+
+// FlowSchemaStatus represents the current state of a flow-schema.
+type FlowSchemaStatus struct {
+	// Current state of flow-schema.
+	Conditions []FlowSchemaCondition `json:"conditions,omitempty" protobuf:"bytes,1,rep,name=conditions"`
+}
+
+// FlowSchemaCondition describes conditions for a flow-schema.
+type FlowSchemaCondition struct {
```

Comment:
A value of this type appears only when every field has a meaningful value.  So I think all the fields are required.  This should be stated in the comments somehow, and not contradicted in the field tags.

Reply:
it's copied/pasted from core/v1 package, and it's consistent w/ most other APIs. let's keep it.

Reply:
We are looking at an evolving code base, some old code is not following best practices.  I am not convinced we should leave this in its current inconsistent state.